oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
-    - 5: unused
-    - 4: unused
-    - 3: unused
-    - 2: unused
-    - 1: end flag. MUST be set in the last frame of a message
-    - 0 (LSB): start flag. MUST be set in the first frame of a message
+  The position of this frame within the overall message. The first frame MUST
+  start at number 0. Each frame's number MUST be 1 higher than the previous
+  frame. Frames for a single message MUST be sent in ascending order of the
+  frame number.
 
-  - `flags[1]`
+- `body_length`, unsigned 32-bit little-endian integer
 
-    - 7 (MSB): unused
-    - 6: unused
-    - 5: unused
-    - 4: unused
-    - 3: unused
-    - 2: unused
-    - 1: unused
-    - 0 (LSB): unused
+  The length of the body byte array of the rfram. This value must not exceed
+  4,080. The sum of the body lengths of all the frames in a message must equal
+  the message length.
 
 - `body`, variable length byte array
 
-  Byte fragment of an encoded message. In order to not exceed the maximum frame
-  length, this byte array MUST NOT exceed 3,936 bytes.
+  Byte fragment of an encoded message. This size of the byte array MUST NOT
+  exceed 4,080 bytes.
+
+The maximum total length of a single frame is 4,096 bytes.
 
 ## Interaction between the Layers
 
 ### Sending Messages
 
-To send a message, it MUST first be encoded into frames. The resulting frame set
-MUST be sent as an uninterrupted sequence.
-
-### Receiving Messages
-
-To receive a message the recipent incrementally reads frames from the
-communication channel. The recipent MAY start parsing a message before all of
-its frames have been received.
-
-### Encoding Messages into Frames
+To send a message, it MUST first be encoded into frames.
 
 The byte representation of a message MUST be chunked into a set of frame bodies.
-The last frame body of the resulting set MAY be shorter than 3,936 bytes (the
-maximum size of a frame body). Any preceding frame bodies MUST be exactly 3,936
-bytes in length.
-
-A frame MUST then be formed from each frame body.
-
-Messages with a length of 3,936 bytes or less MUST be encoded as a single frame.
-
-### Parsing Messages from Frames
-
-Messages are parsed from frames by appending the frame bodies in the order they
-are received.
-
-## Examples for Illustration
-
-### Frame Flags
-
-For messages sent as more than two frames, the first frame would have the start
-flag set, all other flags would be unset. The last frame would have the end flag
-set, all other flags would be unset. Any frames in between would have no flags
-set.
-
-For small messages sent as a single frame, the start flag and the end flag would
-be set, all other flags would be unset.
-
-### Layers Diagram
-
-The diagram below shows a practial example of how the request half of an
-invocation would look like after being encoded as two frames.
+The last frame body of the resulting set MAY be shorter than 4,080 bytes (the
+maximum length of a frame body). Any preceding frame bodies MUST be exactly
+4,080 bytes in length.
```

Comment:
Is this precluding the option of having smaller frames then? That may be problematic if in the future we allow larger frame sizes, since there would be no backwards compatibility. But perhaps we don't care about it to keep things simple.

Reply:
Yes. The alternative is to also specify the expected number of frames. I would rather add a protocol version and fix the frame size per protocol version rather than allowing different frame sizes withing the same protocol.